pkg/grpc/endpoints.go

```diff
@@ -137,11 +134,39 @@ func setAPIEndpointsMiddlewares(rules []config.APIAccessRule) (grpc.UnaryServerI
 		func(srv any, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
 			// Apply the allowlist only on methods that are part of the Dapr runtime, or it will interfere with gRPC proxying
 			if strings.HasPrefix(info.FullMethod, daprRuntimePrefix) {
-				_, ok := allowed[info.FullMethod]
-				if !ok {
-					return invokev1.ErrorFromHTTPResponseCode(http.StatusNotImplemented, "requested endpoint is not available")
+				if len(allowed) > 0 {
+					_, ok := allowed[info.FullMethod]
+					if !ok {
+						return invokev1.ErrorFromHTTPResponseCode(http.StatusNotImplemented, "requested endpoint is not available")
+					}
+				}
+				if len(denied) > 0 {
+					_, ok := denied[info.FullMethod]
+					if ok {
+						return invokev1.ErrorFromHTTPResponseCode(http.StatusNotImplemented, "requested endpoint is not available")
+					}
 				}
 			}
+
 			return handler(srv, stream)
 		}
 }
+
+// Converts a slice of config.APIAccessRule into a map where the key is the gRPC full endpoint
+func apiAccessRuleToMap(rules config.APIAccessRules) map[string]struct{} {
+	res := map[string]struct{}{}
+
+	for _, rule := range rules {
+		if rule.Protocol != "grpc" {
+			continue
+		}
+
+		if list, ok := endpoints[rule.Name+"."+rule.Version]; ok {
```

Comment:
Please use const and fmt, also comment or document it for better maintain.

Reply:
I will document it, but concatenation is significantly faster than fmt.Sprintf for simple string concatenation:

```go
func BenchmarkFmt(b *testing.B) {
	const name = "foo"
	const version = "bar"
	for i := 0; i < b.N; i++ {
		_ = fmt.Sprintf("%s.%s", name, version)
	}
}
func BenchmarkConcat(b *testing.B) {
	const name = "foo"
	const version = "bar"
	for i := 0; i < b.N; i++ {
		_ = name + "." + version
	}
}
```

On my machine:

```text
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr
BenchmarkFmt
BenchmarkFmt-8      	20474700	        58.38 ns/op	       8 B/op	       1 allocs/op
BenchmarkConcat
BenchmarkConcat-8   	1000000000	         0.3238 ns/op	       0 B/op	       0 allocs/op
```

Not just concatenation is MUCH faster (99.95% faster), but fmt.Sprintf also causes heap allocations so more work on the GC.